File: Data Structures & Algorithms/combination-target-sum/submission-6.cpp
class Solution {
private:
    vector<vector<int>> res;
    vector<int> curr;
public:
    void helper(int pos, int sum, vector<int>& nums){
        if (pos == nums.size() || sum < 0){
            if (sum == 0){
                res.push_back(curr);
            }
            return;
        }

        curr.push_back(nums[pos]);
        helper(pos, sum - nums[pos], nums);
        curr.pop_back();

        helper(pos + 1, sum, nums);

    }

    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        helper(0, target, nums);
        return res;
    }
};

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-6.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total = sum(nums)
        target = total // 2

        if total - 2 * target != 0:
            return False
        
        sums = 1

        for num in nums:
            sums |= sums << num

        return (sums >> target) & 1 == 1      